### ISSUE TITLE
Document how to write "aggregate builders"

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/build#926]
i.e. the original request in https://github.com/dart-lang/build/issues/213

/cc @filiph

**Comment by filiph** [comment]
Thanks for filing this!

For convenience, I'm copying over from #213 that we have a good start already — the documentation-by-test provided with https://github.com/dart-lang/build/commit/65b08f6d5710ed78389bf1db74c5dc415e1b4819. 

Please confirm that I'm understanding this right:

* The only magic files are `lib/$lib$`, `web/$web$` and `test/$test$`. This is not generalized to `example/$example$` or `myDir/$myDir$`. Correct?
* This only works when using `pkg:build_runner`. Therefore only useful in `2.0.0-dev`? Correct?
* The only magic that `lib/$lib$` and others provide is that there needs to be no physical input (placeholder) file in these directories. Correct? 
  * If so, how does the build system know to rebuild when I change `data/1.txt`?
  * If not, is the whole of `lib/**` now considered input. In other words, does using `lib/$lib` make the build system rebuild everything on any change to any file in `lib/`?

**Comment by matanlurey** [comment]
@filiph:

> Please confirm that I'm understanding this right:
> 
> * The only magic files are `lib/$lib$`, `web/$web$` and `test/$test$`. This is not generalized to `example/$example$` or `myDir/$myDir$`. Correct?

I believe so. We could add more, though.

_I might misunderstand things, but is there a reason we couldn't support a magic `$package` file?_

/cc @jakemac53 @natebosch 

> * This only works when using `pkg:build_runner`. Therefore only useful in `2.0.0-dev`? Correct?

It works with both `build_runner` and `bazel_codegen`, but does require BR `>=0.7.0`, yes.

> * The only magic that `lib/$lib$` and others provide is that there needs to be no physical input (placeholder) file in these directories. Correct? 

Yes. It creates a "synthetic" asset node. You won't be able to read it as a primary input, but instead you ignore it and just use `listAssets` (or other mechanisms, for example my `ConcatBuilder` example takes configuration instead) to locate and read files.

>   * If so, how does the build system know to rebuild when I change `data/1.txt`?

If you read an asset, it's part of the build graph :)

>   * If not, is the whole of `lib/**` now considered input. In other words, does using `lib/$lib` make the build system rebuild everything on any change to any file in `lib/`?

No.

**Comment by jakemac53** [comment]
> I might misunderstand things, but is there a reason we couldn't support a magic $package file?

Nope, just a design decision to promote good behavior. Technically you can also have `lib/$lib$` as your input extension today and get effectively the same thing.

> If not, is the whole of lib/** now considered input. In other words, does using lib/$lib make the build system rebuild everything on any change to any file in lib/?

We record any globs you pass to `findAssets` and only invalidate on new/deleted nodes that match those globs.

**Comment by matanlurey** [comment]
@jakemac53:

> We record any globs you pass to `findAssets` and only invalidate on new/deleted nodes that match those globs.

We should probably document that (specifically on the API and in our docs)! It could be expensive if folks do a `**` glob, read 2 files, and don't expect this build now to be invalidated frequently.

**Comment by jakemac53** [comment]
> We should probably document that (specifically on the API and in our docs)! It could be expensive if folks do a ** glob, read 2 files, and don't expect this build now to be invalidated frequently.

Yep that is true, at least we only depend on the presence of these files though not their contents.

**Comment by natebosch** [comment]
I think it would be feasible to support `$anyString$` as an extension to match `anyString/` directory - but we'd have to explore how that interacts with the input whitelist.

**Comment by matanlurey** [comment]
Is `$anyString$` more valuable than just a single `$package$`?

**Comment by natebosch** [comment]
> Is `$anyString$` more valuable than just a single `$package$`?

Only if we start adding restrictions around being able to put outputs in different directories by abusing the extension replacement (which IMO we should consider)

**Comment by filiph** [comment]
Thanks all, one more clarification.

> > * This only works when using `pkg:build_runner`. Therefore only useful in `2.0.0-dev`? Correct?
>
> It works with both `build_runner` and `bazel_codegen`, but does require BR `>=0.7.0`, yes.

So adding this to `tool/phases.dart` will not work. Correct?

**Comment by natebosch** [comment]
> So using `lib/$lib$` in something like `tool/phases.dart` will not work. Correct?

It will work as long as you're also on the latest `build_runner`. There were some breaking changes in how your `tool/build.dart` file looks between `0.6.x` and `0.7.x` but it's still a supported path.

**Comment by matanlurey** [comment]
WIP here:
https://github.com/matanlurey/build/blob/147083da9b6a6c70c46eb910a3e046239a2a0a6e/docs/writing_an_aggregate_builder.md

**Comment by filiph** [comment]
Hey, I see Matan's article hasn't landed yet. Will it?

**Comment by matanlurey** [comment]
@filiph: https://github.com/dart-lang/build/blob/master/docs/writing_an_aggregate_builder.md
